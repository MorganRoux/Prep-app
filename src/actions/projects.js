


export const createProject = (project) => ({
    type: 'CREATE_PROJECT',
    project
});

export const removeProject = (id) => ({
    type: 'REMOVE_PROJECT',
    id
});