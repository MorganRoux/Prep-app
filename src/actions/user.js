import { firebase, googleAuthProvider } from '../firebase/firebase';
import database from '../firebase/firebase';
//import project from '../tests/fixtures/projects';

export const startLogin = () => {
    //no need to be here, should stay with loginpage
    return () => {
        return firebase.auth().signInWithPopup(googleAuthProvider);
    }
}

export const startLogout = () => {
    //no need to be here, should stay with equipemntheader
    return () => {
        return firebase.auth().signOut();
    }
}

export const login = (uid) => {
    return {
        type: 'LOGIN',
        uid
    }

}

export const logout = (uid) => {
    return {
        type: 'LOGOUT'
    }
}

export const setUserData = (user) => {
    return {
        type: 'SET_USER_DATA',
        user
    }
}

export const setEquipementRef = (ref) => {
    return {
        type: 'SET_REF',
        ref
    }
}

export const startFetchUserData = (uid) => {
    return (dispatch) => {
        return database.ref(`users/${uid}`)
        .once('value')
        .then( (snapshot) => {

            //test if the user has already logged in
            if( !snapshot.exists() ) {
                throw Error('User not found');
            }

            const {currentProject, profile} = snapshot.val();
            const projectsData = [];
            snapshot.child('/projects').forEach((childSnapshot) => {

                projectsData.push({
                    id: childSnapshot.key,
                    name: childSnapshot.val().name,
                    role: childSnapshot.val().role
                });
            });
            dispatch(setUserData({
                uid,
                currentProject,
                profile,
                projects : projectsData
            }));
        });
    }
}

export const createProfile = (uid) => {
    return (dispatch) => {
        const newProfile = {
            uid,
            currentProject: 'idp1',
            profile: {
                name: 'Anonymous',
                email: 'unknown'
            },
            projects: {
                'idp1' : {
                    name: 'projet1',
                    role: '5'
                }
            }
        };
        return database.ref(`users/${uid}`).set(newProfile)
        .then( () => {
            dispatch(setUserData({
                uid,
                currentProject: 'idp1',
                profile: {
                    name: 'Anonymous',
                    email: 'unknown'
                },
                projects: [{
                    id: 'idp1',
                    name: 'projet1',
                    role: '5'
                }]
            })
        )});
    }
}