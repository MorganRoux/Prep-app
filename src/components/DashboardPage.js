import React from 'react';
import EquipmentList from './EquipmentList';

const DashboardPage = () => {
    return (
        <div>
            <h1>Dashboard Page</h1>
            <EquipmentList />
        </div>
        
    );
};

export default DashboardPage;