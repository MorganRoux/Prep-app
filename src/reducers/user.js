
const userDefaultState = {
    uid: '',
    currentProject: ''
}
const userReducer = (state = userDefaultState, action) => {
    switch(action.type) {
    case 'LOGIN' : 
        return { ...state, uid: action.uid}
    
    case 'LOGOUT' :
        return {...state, uid :''}

    case 'SET_CURRENT_PROJECT' :
        return {...state, currentProject:action.id}
    default:
        return state
    }

}

export default userReducer;