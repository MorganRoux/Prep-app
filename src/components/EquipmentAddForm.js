import React from 'react'
import { Paper, TableBody, TableRow, TableCell, TableFooter } from '@material-ui/core';
import IconButton from '@material-ui/core/IconButton';
import Icon from '@material-ui/core/Icon';

import { startAddEquipment } from '../actions/equipment'

import { connect } from 'react-redux'
import { withSnackbar } from 'notistack'

export class EquipmentAddForm extends React.Component {
    
    constructor(props) {
        super(props);

        this.state = {
            quantity: '',
            publicName: ''
        }
    }

    handleCloseSnackBar = (event, reason) => {
        if (reason === 'clickaway') {
          return;
        }
    
        this.setState({ snackBarOpen: false });
      };

    onQuantityChange = (e) => {
        const quantity = e.target.value;
        if(quantity.match(/^\d*$/)) {
            this.setState( () => ({quantity}));
        }
    }

    onPublicNameChange = (e) => {
        const publicName = e.target.value;
        this.setState( () => ({publicName}));
    }

    onAdd = (e) => {
        e.preventDefault()
        // test the inputs
        if(!this.state.publicName || !this.state.quantity) {
            this.props.enqueueSnackbar('Informations erronnées', {variant:'error'});
            return;
        }

        // search the item in the stocklist
        const itemToAdd = this.props.stocklist.find( (item) => {
            return (item.publicName === this.state.publicName)
        });
        if (!itemToAdd) { 
            this.props.enqueueSnackbar('Élement introuvable', {variant:'error'});
            return;
        }

        // add item
        this.props.startAddEquipment({
            ...itemToAdd,
            quantity: this.state.quantity
        });
        const s = this.state.quantity === '1' ? '':'s'
        this.props.enqueueSnackbar(
            `${this.state.quantity}x ${this.state.publicName} ajouté${s}`, {variant:'success'});

        this.setState( () => ({
            quantity: '',
            publicName:''
        }));
    }

    render () {
        return (
            <TableFooter>
                <TableRow className = "addForm-row">
                    <TableCell />
                    <TableCell colSpan="4">
                        <form onSubmit={this.onAdd}>
                        <input 
                        type="text"
                        name= "quantity"
                        autoFocus
                        placeholder=""
                        value={this.state.quantity}
                        onChange = {this.onQuantityChange}
                        />
                 
                    
                        <input 
                        type="text"
                        name= "publicName"
                        placeholder=""
                        value={this.state.publicName}
                        onChange = {this.onPublicNameChange}
                        />
                    
                        <IconButton className="add-button" aria-label="Add" type="submit">
                            <Icon color="primary">
                            add_circle
                            </Icon>
                        </IconButton>
                        </form>
                    </TableCell>
                </TableRow>
                </TableFooter>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        equipements: state.equipements,
        stocklist: state.stocklist,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startAddEquipment: (item) => dispatch(startAddEquipment(item))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(withSnackbar(EquipmentAddForm))