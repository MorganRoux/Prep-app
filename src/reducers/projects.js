import projects from '../tests/fixtures/projects';

const projectsDefaultState = projects;

const projectsReducer = ( state = projectsDefaultState, action ) => {

    switch(action.type) {
    case 'CREATE_PROJECT':
        return state ? ([
            ...state,
            {
                ...action.project,
                role:'5'
            }
        ]) : ([{
            ...action.project,
            role:'5'
        }]);
        

    case 'REMOVE_PROJECT' :
        return state.filter( (project) => (project.id !== action.id));

    default:
        return state;
    }
}

export default projectsReducer