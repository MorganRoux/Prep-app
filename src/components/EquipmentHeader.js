import React from 'react'
import { connect } from 'react-redux';
import InputBase from '@material-ui/core/InputBase';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import { setTextFilter } from '../actions/filters';
import { startLogout } from '../actions/user';

export class EquipmentHeader extends React.Component {

    constructor (props) {
        super(props);
        this.state = {
            searchText : ''
        }
    }

    onSearchChange = (e) => {

        const searchText = e.target.value;
        this.setState(() => ({searchText}));
        this.props.setTextFilter(searchText);
    }
    
    render () {
        return (
            <div style = {{display: 'flex', justifyContent: 'space-between'}}>
                <div style = {{display: 'flex', justifyContent: 'flex-start'}}>
                    <IconButton>
                        <MenuIcon />
                    </IconButton>
                    <h1>Equipment List</h1>
                </div>
                <div style = {{display: 'flex', justifyContent: 'flex-end'}}>
                    <InputBase 
                        className="search-field"
                        placeholder="Rechercher" 
                        value={this.state.searchText}
                        onChange={this.onSearchChange}
                    />
                    <IconButton>
                        <SearchIcon />
                    </IconButton>
                    <button className = "logout-button" onClick={this.props.startLogout}>Logout</button>
                </div>
            </div>
    )};

}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter : (text) =>  dispatch(setTextFilter(text)),
    startLogout : () => dispatch(startLogout())
});

export default connect(null, mapDispatchToProps)(EquipmentHeader)