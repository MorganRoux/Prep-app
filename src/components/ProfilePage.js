import React from 'react';
import { connect } from 'react-redux';
import { setCurrentProject } from '../actions/user';
import Select from '@material-ui/core/Select';

export class ProfilePage extends React.Component { 

    onChange = (e) => {
        this.props.setCurrentProject(e.target.value);
    }

    render() {
        console.log(this.props.projects);
        return (
            <div>
                <h1>ProfilePage</h1>

                
                    <Select
                        native
                        onChange={this.onChange}
                    >
                    { this.props.projects.map( (project) => {
                        <option value={project.id}>{project.name}</option>
                    })}
                        
                    </Select>
            </div>
        );
    }
    
};

const mapStateToProps = (state) => ({
    user: state.user,
    projects: state.projects
});

const mapDispatchToProps = (dispatch) => ({
    setCurrentProject: (id) => dispatch(setCurrentProject(id))
})
export default connect(mapStateToProps,mapDispatchToProps)(ProfilePage);