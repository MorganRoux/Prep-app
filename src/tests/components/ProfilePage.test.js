import React from 'react';
import { ProfilePage } from '../../components/ProfilePage';
import user from '../fixtures/user';
import projects from '../fixtures/projects'
import { shallow } from 'enzyme';
import Select from '@material-ui/core/Select';

test('should render profile page', () => {

    const setCurrentProject = jest.fn();
    const wrapper = shallow(
        <ProfilePage 
            setCurrentProject={setCurrentProject}
            user={user}
            projects={projects}
        />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle select change', () => {

    const setCurrentProject = jest.fn();
    const id='1';
    const wrapper = shallow(
        <ProfilePage 
            setCurrentProject={setCurrentProject}
            user={user}
            projects={projects}
        />);
    wrapper.find(Select).simulate('change', {target:{value: id}});
    expect(setCurrentProject).toHaveBeenCalledWith(id);

});

