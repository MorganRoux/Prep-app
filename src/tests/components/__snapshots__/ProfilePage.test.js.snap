// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render profile page 1`] = `
<div>
  <h1>
    ProfilePage
  </h1>
  <WithStyles(WithFormControlContext(Select))
    native={true}
    onChange={[Function]}
  />
</div>
`;
